node/src/components/linear_chain_sync.rs

```diff
@@ -26,114 +52,351 @@ pub trait ReactorEventT<I>:
+                    info!(%block_height, "Finished synchronizing linear chain up until trusted hash.");
+                    let peer = self.random_peer_unsafe();
+                    // Kick off syncing trusted hash descendants.
+                    self.state = State::sync_descendants(trusted_hash);
+                    fetch_block_at_height(effect_builder, peer, block_height + 1)
+                } else {
+                    self.state = curr_state;
+                    self.fetch_next_block_deploys(effect_builder)
+                }
+            }
+            State::SyncingDescendants {
+                ref current_block, ..
+            } => {
+                match current_block.as_ref() {
+                    Some(expected) => assert_eq!(
+                        expected, &block_header,
+                        "Block execution result doesn't match received block."
+                    ),
+                    None => panic!("Unexpected block execution results."),
+                }
+                self.state = curr_state;
+                self.fetch_next_block(effect_builder, rng, &block_header)
             }
-            Some(block) => fetch_block_deploys(effect_builder, peer, block),
         }
     }
 
-    pub(crate) fn init_block_era(&self) -> Option<EraId> {
-        self.init_block_era
+    /// Returns effects for fetching next block's deploys.
+    fn fetch_next_block_deploys<REv>(
+        &mut self,
+        effect_builder: EffectBuilder<REv>,
+    ) -> Effects<Event<I>>
+    where
+        I: Send + Copy + 'static,
+        REv: ReactorEventT<I>,
+    {
+        let peer = self.random_peer_unsafe();
+
+        let next_block = match self.state {
+            State::None | State::Done => {
+                panic!("Tried fetching next block when in {:?} state.", self.state)
+            }
+            State::SyncingTrustedHash {
+                ref mut linear_chain,
+                ref mut current_block,
+                ..
+            } => match linear_chain.pop() {
+                None => None,
+                Some(block) => {
+                    // Update `current_block` so that we can verify whether result of execution
+                    // matches the expected value.
+                    current_block.replace(block.clone());
+                    Some(block)
+                }
+            },
+            State::SyncingDescendants {
+                ref mut linear_chain_block,
+                ref mut current_block,
+                ..
+            } => match linear_chain_block.take() {
+                None => None,
+                Some(block) => {
+                    // Update `current_block` so that we can verify whether result of execution
+                    // matches the expected value.
+                    current_block.replace(block.clone());
+                    Some(block)
+                }
+            },
+        };
+
+        next_block.map_or_else(Effects::new, |block| {
```

Comment:
I've followed the flow, and it seems like wherever this is called from, next block should never be `None`.  Maybe we should be panicking here, or at least logging an error, since that would stall the flow I think?

Reply:
Added a warning for now as I am not 100% sure whether it's really an error that should crash the node or just a harmless bug.